### PR TITLE
Show nicer error when a @ConfigMapping class exists in the deployment module and set to the runtime phase

### LINKED ISSUE #43394: Show nicer error when a `@ConfigMapping` class exists in the `deployment` module and set to the `runtime` phase
### Description

While creating (see this PR for more context - https://github.com/quarkusio/quarkus-workshops/pull/600), I noticed that the following error is thrown when the `@ConfigMapping` phase is set to RUN_TIME:

```java
Caused by: java.lang.ClassNotFoundException: io.quarkus.workshop.superheroes.version.deployment.VersionConfig
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:533)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:533)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:481)
	at io.quarkus.runtime.configuration.AbstractConfigBuilder.withMapping(AbstractConfigBuilder.java:64)
	... 21 more

``` 

The solution was to move the `@ConfigMapping` to the `runtime` module, but the error message is a bit cryptic.

- Zulip thread: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Workshop/near/471292983

### Implementation ideas

It would be nice if we could detect that somehow and display a nicer error message

### PR DESCRIPTION
**Pull request by mcruzdev**
- Fixes: #43394

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java [hunk 1]
```diff
@@ -631,6 +631,19 @@ private static void generateConfigBuilder(
             }
 
             for (ConfigClassWithPrefix mapping : mappings) {
+                boolean isPresentAtDeployment = isClassAvailable(deploymentCl,
```
Comment (radcortez) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java:634]: This should be checked in https://github.com/quarkusio/quarkus/blob/be4817d541882225767e9e3655ec5bf325045360/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java#L98-L111.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java [hunk 2]
```diff
@@ -631,6 +631,19 @@ private static void generateConfigBuilder(
...
+                        mapping.getKlass().getName());
+
+                // necessary because core/runtime configurations does no exist on QuarkusClassLoader.
+                if (!mapping.getKlass().getName().startsWith("io.quarkus.runtime.")) {
```
Comment (radcortez) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java:638]: Wouldn't it be easier to check, if mapping is `RUNTIME` and `QuarkusClassLoader.isClassPresentAtRuntime`? We may want to include `BUILD_AND_RUNTIME_FIXED` also.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java (hunk 1 above)
Comment (radcortez) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java:634]: Or better yet in here:
https://github.com/quarkusio/quarkus/blob/be4817d541882225767e9e3655ec5bf325045360/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java#L166-L184

Which already reports other checks.

**Review by radcortez — CHANGES_REQUESTED** [review]

**Comment by radcortez** [comment]
Thank you for the PR; I've left a few comments.

**Review by radcortez — CHANGES_REQUESTED** [review]
The CI failure is related.

**Review by radcortez — CHANGES_REQUESTED** [review]
Can we add a test please? Thank you.

File: core/deployment/src/test/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReaderTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+import io.smallrye.config.ConfigMapping;
+
+class BuildTimeConfigurationReaderTest {
```
Comment (radcortez) [inline core/deployment/src/test/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReaderTest.java:14]: My recommendation is to add the test here: https://github.com/quarkusio/quarkus/tree/4e11ab3c285ff7568c693d47a515122a3e7503d4/integration-tests/test-extension/extension

And use a `QuarkusUnitTest`.

Comment (radcortez) [reply core/deployment/src/test/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReaderTest.java:14]: What was the issue?

Comment (radcortez) [reply core/deployment/src/test/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReaderTest.java:14]: You can selectively create an inner class with the mapping local to the test class and then manually register it with:

```java
.addAsResource(new StringAsset("MAPPING_NAME"), "META-INF/quarkus-config-roots.list")
```

It should only affect the specific test and shouldn't leak to other tests.

**Review by radcortez — APPROVED** [review]

**Comment by radcortez** [comment]
It seems that the CL check is not reliable at all. There is already a workaround for the configs in the `core` module, and now there are some related test failures. 

Maybe we should implement this in the annotation processor instead. We don't have a way to determine what kind of module we are analysing, but we can probably come up with some ideas.

**Review by radcortez — CHANGES_REQUESTED** [review]
Test failures are related.

**Comment by gsmet** [comment]
I agree that maybe doing it in the extension annotation processor would be easier. I have some logic to detect runtime/deployment that I need to clean up anyway.

I'll have a look at making it more reliable and you could then use that instead for your patch.

**Comment by mcruzdev** [comment]
Hi @gsmet and @radcortez I think it is ok now, could take a new review?

**Comment by gsmet** [comment]
Looks like CI is passing! Could you squash the commits?

I'm traveling for 2 days, will have a look after that if noone beats me to it.

**Comment by radcortez** [comment]
> I'm traveling for 2 days, will have a look after that if noone beats me to it.

Don't worry about it. I'll have a look.

**Comment by mcruzdev** [comment]
> Looks like CI is passing! Could you squash the commits?

Done!

**Review by radcortez — APPROVED** [review]

**Comment by radcortez** [comment]
I gave it a try, and it worked as expected. I think this last implementation is way better than what we had before.

Thank you for the work and patience @mcruzdev

File: core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/AbstractConfigListener.java [hunk 4]
```diff
@@ -66,6 +68,16 @@ public void onResolvedEnum(TypeElement enumTypeElement) {
         configCollector.addResolvedEnum(enumDefinition);
     }
 
+    protected void validateRuntimeConfigOnDeploymentModules(ConfigPhase configPhase) {
+        if (configPhase.equals(ConfigPhase.RUN_TIME) || configPhase.equals(ConfigPhase.BUILD_AND_RUN_TIME_FIXED)) {
+            ExtensionModule.ExtensionModuleType type = config.getExtensionModule().type();
+            if (type.equals(ExtensionModule.ExtensionModuleType.DEPLOYMENT)) {
+                throw new IllegalStateException("Configuration classes with ConfigPhase.RUN_TIME or " +
+                        "ConfigPhase.BUILD_AND_RUNTIME_FIXED phases, must be on runtime module.");
```
Comment (gastaldi) [inline core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/AbstractConfigListener.java:76]:
```suggestion
                        "ConfigPhase.BUILD_AND_RUNTIME_FIXED phases must be on runtime module.");
```

File: core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/ConfigMappingListener.java [hunk 5]
```diff
@@ -75,6 +75,9 @@ public Optional<DiscoveryConfigRoot> onConfigRoot(TypeElement configRoot) {
             }
         }
 
+        validateRuntimeConfigOnDeploymentModules(
+                configPhase);
```
Comment (gastaldi) [inline core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/ConfigMappingListener.java:79]:
```suggestion
        validateRuntimeConfigOnDeploymentModules(configPhase);
```

File: core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/AbstractConfigListener.java (hunk 4 above)
Comment (gastaldi) [reply core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/AbstractConfigListener.java:76]: Also I think we should log the offending `ConfigMapping` class name in the message

File: core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/ConfigMappingListener.java (hunk 5 above)
Comment (mcruzdev) [reply core/processor/src/main/java/io/quarkus/annotation/processor/documentation/config/scanner/ConfigMappingListener.java]: Thank you GOAT! suggestions applied :)

**Review by gastaldi — APPROVED** [review]

**Pull request merged**